pkg/quota/image/usage.go

```diff
@@ -10,139 +13,400 @@ import (
+			ns = objRef.Namespace
+		}
+
+		isName := nameParts[0]
+		is, err := c.getImageStream(ns, isName)
+		if err != nil {
+			return imageapi.DockerImageReference{}, fmt.Errorf("failed to get imageStream for ImageStreamTag %s/%s: %v", ns, objRef.Name, err)
+		}
+
+		event := imageapi.LatestTaggedImage(is, nameParts[1])
+		if event == nil || len(event.DockerImageReference) == 0 {
+			return imageapi.DockerImageReference{}, fmt.Errorf("%q is not currently pointing to an image, cannot use it as the source of a tag", objRef.Name)
+		}
+		return imageapi.ParseDockerImageReference(event.DockerImageReference)
+
+	case "DockerImage":
+		managedByOS, ref := imageReferenceBelongsToInternalRegistry(objRef.Name)
+		if !managedByOS {
+			return imageapi.DockerImageReference{}, fmt.Errorf("DockerImage %s does not belong to internal registry", objRef.Name)
 		}
+		return ref, nil
 	}
 
-	return
+	return imageapi.DockerImageReference{}, fmt.Errorf("unsupported object reference kind %s", objRef.Kind)
+}
+
+// getImageStream gets an image stream object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImageStream(namespace, name string) (*imageapi.ImageStream, error) {
+	key := fmt.Sprintf("%s/%s", namespace, name)
+	if is, exists := c.imageStreamCache[key]; exists {
+		return is, nil
+	}
+	is, err := c.osClient.ImageStreams(namespace).Get(name)
+	if err == nil {
+		c.imageStreamCache[key] = is
+	}
+	return is, err
+}
+
+// getImage gets image object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImage(name string) (*imageapi.Image, error) {
+	if image, exists := c.imageCache[name]; exists {
+		return image, nil
+	}
+	image, err := c.osClient.Images().Get(name)
+	if err == nil {
+		c.imageCache[name] = image
+	}
+	return image, err
+}
+
+// getImageStreamImage gets an image belonging to a given image stream object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImageStreamImage(namespace, name string) (*imageapi.Image, error) {
+	nameParts := strings.SplitN(name, "@", 2)
+	if len(nameParts) != 2 {
+		return nil, fmt.Errorf("failed to parse image stream image name %q, expected exactly one '@'", name)
+	}
+	image, exists := c.imageCache[nameParts[1]]
+	if exists {
+		return image, nil
+	}
+	isi, err := c.osClient.ImageStreamImages(namespace).Get(nameParts[0], nameParts[1])
+	if err == nil {
+		image = &isi.Image
+		c.imageCache[name] = image
+	}
+	return image, err
+}
+
+// listImageStreams returns a list of image streams of the given namespace and caches them for later access.
+func (c *GenericImageStreamUsageComputer) listImageStreams(namespace string) (*imageapi.ImageStreamList, error) {
+	iss, err := c.osClient.ImageStreams(namespace).List(kapi.ListOptions{})
+	if err == nil {
+		for _, is := range iss.Items {
+			c.imageStreamCache[fmt.Sprintf("%s/%s", namespace, is.Name)] = &is
```

Comment:
This cache is always thrown away after use. It's recreated on each admission check and resource quota sync. The admission needs to work with the most up to date object available and don't attempt to update it until the check is finished. I don't think this use case allows for sharing cache with the controller.